### PR TITLE
fix(api): harden abuse-gate error logging

### PR DESCRIPTION
**Pull request by DysektAI**
## Summary

Hardens the API abuse gate so binding exceptions cannot disclose the raw client IP in structured Worker logs, and aligns the public/operational rate-limit documentation with the fail-open daily-quota behavior.

## Changes

- Redact the current `CF-Connecting-IP` value from abuse-gate exception names and messages before logging.
- Add regression coverage for IP redaction and make rate-limit test cleanup/payload timing deterministic.
- Remove obsolete rate-limiter `503` responses from the OpenAPI contract.
- Clarify when daily-quota headers are present and keep `IP_HASH_SECRET` guidance synchronized.
- Correct legacy Durable Object retention guidance and track staged cleanup in #594.

## Related Issues

Follow-up to #591. Related to #594.

## Testing

- [x] Tested locally
- [x] Tested in production-like environment
- [x] Added/updated unit tests
- [x] Manual testing performed

### Test Plan

1. Ran `pnpm run test:api-gateway` (7 files, 106 tests), `pnpm run test` (218 files, 2,160 tests), `pnpm run typecheck`, and `pnpm run lint`.
2. Ran `pnpm run validate:openapi`, targeted Prettier checks, and an API Gateway Wrangler deployment dry run.
3. Provisioned and verified the production `IP_HASH_SECRET`, then smoke-tested the production health and OpenAPI endpoints.

## Screenshots/Videos

Not applicable; no UI changes.

## Documentation impact

- [ ] No behavior changed — no documentation review needed
- [ ] Behavior changed — existing documentation remains accurate
- [x] Behavior changed — documentation was updated in this PR
- [ ] Behavior changed — follow-up documentation issue: #

Documents reviewed when relevant:

- [ ] README or user guidance
- [x] Contributor documentation
- [ ] SYSTEMS or architecture
- [x] API or rate-limit reference
- [x] Runbook or environment variables
- [ ] Screenshots or diagrams

## Checklist

- [x] My code follows the project's coding conventions (see AGENTS.md)
- [x] I have performed a self-review of my code
- [x] I have commented my code where necessary
- [x] My changes generate no new warnings or errors
- [x] I have tested my changes thoroughly
- [x] All existing tests still pass
- [x] I have checked for any breaking changes

## Additional Notes

PR #595 was superseded because its branch retained the pre-squash #591 history and conflicted with `main`. This branch is based directly on the merged #591 commit and contains only the follow-up fixes.

The Kilo review on #595 failed because its agent wrapper failed before producing findings; its #596 rerun completed successfully. CodeRabbit reviewed the original patch on #595 but rate-limited the #596 rerun. A direct self-review of the complete #596 diff, including the latest review fixes, found no additional correctness, security, privacy, or operational issues. All available feedback was reconciled.

The current link-check failure is caused by pre-existing GitHub Discussions links in `README.md` and `SUPPORT.md`; the dedicated, mergeable fix is #593. This PR does not duplicate that unrelated documentation change.

<a href="https://app.devin.ai/review/tarkovtracker-org/tarkovtracker/pull/596" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

<a href="https://cubic.dev/pr/tarkovtracker-org/TarkovTracker/pull/596?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Documentation**
  * Clarified rate-limit headers and `429` differences between daily-quota vs abuse-gate denials, including fail-open behavior.
  * Updated legacy cleanup/retention guidance plus related architecture/runbook wording.
* **API Changes**
  * Removed documented `503 Service Unavailable` responses from multiple endpoints in the OpenAPI spec.
* **Bug Fixes**
  * Improved rate-limiter error logging to safely format error details and redact sensitive values (including IPv6 cases).
* **Tests**
  * Expanded and streamlined rate-limit tests for abuse-gate/daily-quota edge cases and logging behavior.

___

## **CodeAnt-AI Description**
Protect abuse-gate logs from leaking client IP addresses

### What Changed
- Abuse-gate error logs now redact client IPv4 and mixed-case IPv6 addresses, including when the rate-limit service changes the address casing.
- Logging remains safe when an error’s name or message cannot be read, while requests continue through the existing fail-open behavior.
- API documentation no longer advertises unsupported `503` rate-limit responses and clarifies when quota headers are returned.
- Updated retention, secret, and rate-limit guidance to match current behavior, with regression coverage for redaction and error handling.

### Impact
`✅ No raw client IPs in abuse-gate error logs`
`✅ Requests continue when abuse-rate limiting is unavailable`
`✅ API rate-limit documentation matches actual responses`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

<details open><summary><h3>Greptile Summary</h3></summary>

This PR hardens the API abuse-gate error-logging path to prevent client IP addresses from appearing in Worker logs, and cleans up supporting documentation and the OpenAPI spec.

- **IP redaction in error logs**: Introduces `sanitizeLogValue` and `readErrorProperty` helpers that apply case-insensitive redaction of the client IP from any exception `name`/`message` before logging, with a 256-char intermediate cap to ensure IPs straddling the final 200-char truncation boundary are fully replaced before trimming.
- **Test expansion**: Adds regression tests for IPv6 case-normalization, boundary straddling, and getter-throwing errors; consolidates `warnSpy.mockRestore()` calls into a single `vi.restoreAllMocks()` in `afterEach`; and converts `malformedQuotaCases` payloads to factory functions to eliminate stale `Date.now()` evaluation.
- **Spec/doc cleanup**: Removes documented `503` responses (gateway is fail-open), updates `X-RateLimit-*` header presence description, corrects legacy DO retention guidance, and refines `IP_HASH_SECRET` runbook wording.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — changes are narrowly scoped to logging helpers, test cleanup, and documentation; no request-handling or auth logic is altered.

The redaction helpers are well-bounded: each path through `sanitizeLogValue` and `readErrorProperty` has its own exception guard, the call site threads a non-null `clientIp` through consistently, and the 256-char intermediate cap correctly ensures IPs straddling the 200-char truncation boundary are fully replaced before trimming. Tests cover IPv6 case normalization, boundary straddling, and getter-throwing errors. Documentation and OpenAPI changes accurately reflect existing runtime behavior.

**Files Needing Attention:** No files require special attention.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| workers/api-gateway/src/index.ts | Adds `sanitizeLogValue`, `readErrorProperty`, and updates `errorInfo` to accept a `redactions` array; call site correctly threads `clientIp` through; logic handles empty redactions, getter exceptions, and length guard correctly. |
| workers/api-gateway/src/__tests__/rate-limits.test.ts | New tests cover IPv6 case-insensitive redaction, boundary-straddling truncation, and getter-throwing errors; `malformedQuotaCases` factory-function fix prevents stale `Date.now()` in timing-sensitive cases; mock cleanup consolidated into `afterEach`. |
| workers/api-gateway/src/openapi.ts | Removes `ServiceUnavailable` (503) component and its references from all seven endpoints; accurately reflects the fail-open behavior where no 503 is ever returned. |
| docs/API.md | Clarifies that `X-RateLimit-*` headers are present only when the daily-quota service returns a quota decision, accurately reflecting fail-open behavior. |
| docs/RATE_LIMITING.md | Corrects legacy DO retention row to reflect that gateway-created retained keys did not schedule cleanup alarms and will persist until the staged #594 cleanup. |
| docs/ARCHITECTURE.md | Minor wording improvement to `IP_HASH_SECRET` description; broadens "429 log lines" to "abuse-gate log lines" to match actual behavior. |
| docs/runbook.md | Narrows pre-deploy secret-check description to mention `ip_hash` specifically, improving clarity for operators. |

</details>

<sub>Reviews (7): Last reviewed commit: ["fix(api): correct boundary redaction tes..."](https://github.com/tarkovtracker-org/tarkovtracker/commit/3f19c68bf3e80f128784c850d8d4d943917e7d54) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=47128441)</sub>

### CODE REVIEW

File: workers/api-gateway/src/index.ts [hunk 1]
```diff
@@ -492,11 +492,32 @@ async function hashIp(ip: string, secret?: string): Promise<string | null> {
...
+      error_name: sanitizeLogValue(error.name, redactions),
+      error_message: sanitizeLogValue(error.message, redactions),
+    };
   }
-  return { error_name: typeof error, error_message: String(error).slice(0, 200) };
+  return {
+    error_name: sanitizeLogValue(typeof error, redactions),
+    error_message: sanitizeLogValue(error, redactions),
+  };
 }
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts:541]: Confirmed. Redaction runs before truncation and uses the exact IP literal used in the abuse-gate key, so the binding-error leak path is covered. No additional change needed.

File: workers/api-gateway/src/openapi.ts [hunk 2]
```diff
@@ -160,19 +160,6 @@ export const OPENAPI_SPEC = {
...
-          'application/json': {
-            schema: { $ref: '#/components/schemas/ErrorResponse' },
-            examples: {
-              limiterUnavailable: {
-                value: { success: false, error: 'Rate limiter unavailable' },
-              },
-            },
-          },
-        },
-      },
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/openapi.ts:175]: Confirmed. Runtime daily-quota failures are deliberately fail-open, the Worker has no rate-limiter 503 path, and `pnpm run validate:openapi` passes with no dangling component references. No additional change needed.

File: workers/api-gateway/src/__tests__/rate-limits.test.ts [hunk 3]
```diff
@@ -637,51 +638,58 @@ describe('daily quota and abuse gate', () => {
...
       name: 'contradictory allowed state',
-      payload: { allowed: false, remaining: 3, resetAt: Date.now() + 1000 },
+      payload: () => ({ allowed: false, remaining: 3, resetAt: Date.now() + 60_000 }),
     },
     {
       name: 'far-future resetAt',
-      payload: { allowed: false, remaining: 0, resetAt: Date.now() + 86400000 * 2 },
+      payload: () => ({ allowed: false, remaining: 0, resetAt: Date.now() + 86_400_000 * 2 }),
     },
   ];
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/__tests__/rate-limits.test.ts:808]: Confirmed. Payload factories now evaluate timestamps inside each test, and `afterEach` guarantees mock restoration even after failed assertions. The API Gateway suite passes with 108 tests.

File: workers/api-gateway/src/index.ts [hunk 4]
```diff
@@ -492,11 +492,32 @@ async function hashIp(ip: string, secret?: string): Promise<string | null> {
...
+function sanitizeLogValue(value: unknown, redactions: string[]): string {
+  let sanitized: string;
+  try {
+    sanitized = String(value);
+  } catch {
+    return 'Unknown';
+  }
+  for (const redaction of redactions) {
+    sanitized = sanitized.replaceAll(redaction, '[redacted]');
+  }
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts:513]: Valid and fixed in `887125fd`: empty redaction values are skipped inside `sanitizeLogValue`, making the helper safe independently of its current guarded call site.

File: workers/api-gateway/src/index.ts [hunk 5]
```diff
@@ -492,11 +492,32 @@ async function hashIp(ip: string, secret?: string): Promise<string | null> {
...
+}
+function errorInfo(
+  error: unknown,
+  redactions: string[]
+): { error_name: string; error_message: string } {
   if (error instanceof Error) {
-    return { error_name: error.name, error_message: error.message.slice(0, 200) };
+    return {
+      error_name: sanitizeLogValue(error.name, redactions),
+      error_message: sanitizeLogValue(error.message, redactions),
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts]: Valid and fixed in `887125fd`: Error property reads are isolated behind a guarded accessor, `instanceof` failures also fall back safely, and integration tests verify fail-open behavior when either `name` or `message` throws.

File: docs/API.md [hunk 6]
```diff
@@ -393,7 +393,7 @@ A pre-authentication IP-based abuse gate (Cloudflare Workers Rate Limiting bindi
...
+Responses for which the daily-quota service returns a quota decision include `X-RateLimit-Limit`, `X-RateLimit-Remaining`, and `X-RateLimit-Reset` (Unix seconds), plus `Retry-After` when that quota returns `429`. Pre-authentication abuse-gate `429` responses include only `Retry-After`, and fail-open responses (daily-quota service temporarily unavailable) omit the `X-RateLimit-*` headers. When a free-tier user exhausts a daily quota, the `429` body includes an upgrade link. Admins can inspect the top consumers via `GET /api/admin/api-usage`; usage is bucketed by UTC day, so the report covers the current and previous UTC day (the `since` field gives the exact starting day).
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply docs/API.md]: Valid and fixed in `887125fd`: the docs now state that the quota service returns a decision and the gateway converts a deny decision into the client-facing 429 with `Retry-After`.

**Comment by DysektAI** [comment]
@coderabbitai Full Review

**Comment by DysektAI** [comment]
Valid concern — addressed in `de3108eb`.

`sanitizeLogValue` previously used case-sensitive `replaceAll`, so a rate-limit binding that lowercased an IPv6 address's hex digits before echoing it in an exception would leave the normalized form in `error_message`. The fix switches to case-insensitive regex replacement (with regex-special-char escaping to prevent injection from the IP value):

```ts
const escaped = redaction.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
sanitized = sanitized.replace(new RegExp(escaped, 'gi'), '[redacted]');
```

Added a regression test (`redacts a mixed-case IPv6 address even when the binding lowercases it`) that sends `CF-Connecting-IP: 2001:DB8::1`, simulates a binding error containing the lowercased `2001:db8::1`, and asserts neither form appears in the log. All 109 api-gateway tests pass.

**Comment by DysektAI** [comment]
Valid concern — addressed in `9cdb358c`.

`sanitizeLogValue` now caps the error string to 200 chars before constructing or executing the redaction regex, and skips redactions longer than the capped string (they can't match). This bounds the CPU/memory work to a fixed 200-char window regardless of the input size:

```ts
sanitized = sanitized.slice(0, 200);
for (const redaction of redactions) {
  if (redaction.length === 0 || redaction.length > sanitized.length) continue;
  // ...
}
```

All 109 api-gateway tests pass.

File: workers/api-gateway/src/index.ts [hunk 7]
```diff
@@ -492,11 +492,51 @@ async function hashIp(ip: string, secret?: string): Promise<string | null> {
...
+function sanitizeLogValue(value: unknown, redactions: string[]): string {
+  let sanitized: string;
+  try {
+    sanitized = String(value);
+  } catch {
+    return 'Unknown';
+  }
+  // Cap early: this runs pre-auth on attacker-influenced inputs, so avoid
+  // unbounded string/regex work before the final truncation.
+  sanitized = sanitized.slice(0, 200);
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts]: Valid concern — addressed in `e27f7afe`.

The early `slice(0, 200)` could split an IP across the truncation boundary, leaving a fragment when the full pattern failed to match. The fix caps to 256 (200 + max IPv6 length of 45 with margin) before redaction, so any IP starting before char 200 is fully present for replacement, then slices to 200 after:

```ts
sanitized = sanitized.slice(0, 256);
for (const redaction of redactions) {
  if (redaction.length === 0 || redaction.length > sanitized.length) continue;
  const escaped = redaction.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  sanitized = sanitized.replace(new RegExp(escaped, "gi"), "[redacted]");
}
return sanitized.slice(0, 200);
```

This preserves the bounded-input guarantee (256 chars max) while eliminating the partial-IP leak window. Added a boundary regression test (`redacts an IP that straddles the 200-char truncation boundary`) that places an IP at char 195 and asserts no fragment appears. All 110 api-gateway tests pass.

File: workers/api-gateway/src/__tests__/rate-limits.test.ts [hunk 8]
```diff
@@ -637,51 +638,181 @@ describe('daily quota and abuse gate', () => {
...
+    });
+    expect(warnLog).not.toContain('2001:db8::1');
+    expect(warnLog).not.toContain('2001:DB8::1');
+  });
+  it('redacts an IP that straddles the 200-char truncation boundary', async () => {
+    const warnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    const calls: LimiterCall[] = [];
+    const ip = '203.0.113.99';
+    // Place the IP so it starts before char 200 but extends past it.
+    const padding = 'x'.repeat(195);
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/__tests__/rate-limits.test.ts]: Valid — addressed in `3f19c68b`.

The previous padding (195) placed the IP at char 219, fully past the 200 boundary. The test passed by truncation, not by exercising boundary redaction. Fixed by deriving padding from the prefix length so the IP starts at char 190 and straddles 200, and added `expect(warnLog).toContain('[redacted]')` to assert the redaction actually fired. All 110 api-gateway tests pass.
